### PR TITLE
Feature/us 391967 - Hazelcast environment variables to infinity

### PR DESCRIPTION
**Pull request by Saurabh-16**
Changes for adding hazelcast configurations in pega web ready docker image.
We need to pass few settings to pega infinity to ensure hazelcast runs in client server mode.

### CODE REVIEW

File: tomcat-conf/Catalina/localhost/prweb.xml [hunk 1]
```diff
@@ -32,5 +32,13 @@
   <Environment name="prconfig/session/ha/enabled" value="true" type="java.lang.String"/>
   <Environment name="prconfig/dsm/services/stream/pyBaseLogPath" value="/opt/pega/kafkadata" type="java.lang.String" />
   <Environment name="prconfig/cloud/isNodeAgnosticAdminStudio" value="true" type="java.lang.String"/>
+
+  {{ if isTrue .Env.CLIENT_MODE }}
+  <Environment name="prconfig/cluster/clientserver/clientmode" value="{{ .Env.CLIENT_MODE }}" type="java.lang.String"/>
+  <Environment name="prconfig/Pega-RULES/onPegaK8s" value="{{ .Env.ON_PEGA_K8S }}" type="java.lang.String"/>
```
Comment (dcasavant) [inline tomcat-conf/Catalina/localhost/prweb.xml:38]: Why do we need to have a parameter that specifies architecture implementation?  Should we be naming this parameter based on the action it performs instead?

Also, this Docker image is only used on Kubernetes deployments today.

File: tomcat-conf/Catalina/localhost/prweb.xml [hunk 2]
```diff
@@ -32,5 +32,13 @@
...
+  <Environment name="prconfig/identification/cluster/name" value="{{ .Env.CLUSTER_NAME }}" type="java.lang.String"/>
+  <Environment name="prconfig/cluster/clientserver/server/namespace" value="{{ .Env.NAMESPACE }}" type="java.lang.String"/>
```
Comment (dcasavant) [inline tomcat-conf/Catalina/localhost/prweb.xml:40]: Instead of server namespace and service name, if we just used URL, wouldn't it cover both?

File: tomcat-conf/Catalina/localhost/prweb.xml [hunk 3]
```diff
@@ -32,5 +32,13 @@
   <Environment name="prconfig/session/ha/enabled" value="true" type="java.lang.String"/>
   <Environment name="prconfig/dsm/services/stream/pyBaseLogPath" value="/opt/pega/kafkadata" type="java.lang.String" />
   <Environment name="prconfig/cloud/isNodeAgnosticAdminStudio" value="true" type="java.lang.String"/>
+
+  {{ if isTrue .Env.CLIENT_MODE }}
```
Comment (dcasavant) [inline tomcat-conf/Catalina/localhost/prweb.xml:36]: As a parameter, `CLIENT_MODE` sounds like this node will be accepting clients IMO.  Can we go with something more descriptive that is user facing? We had previously called this the cluster service, correct?Maybe something like `CLUSTER_SERVICE_URL` or `DATA_GRID_URL`. We could infer we should enable this mode if a URL is provided.  Thoughts?

File: tomcat-conf/Catalina/localhost/prweb.xml (hunk 2 above)
Comment (Saurabh-16) [reply tomcat-conf/Catalina/localhost/prweb.xml:40]: Will make this change.

File: tomcat-conf/Catalina/localhost/prweb.xml (hunk 3 above)
Comment (Saurabh-16) [reply tomcat-conf/Catalina/localhost/prweb.xml:36]: Ok, instead of checking CLIENT_MODE as true , I could infer this based on whether URL is provided or not. It will then look something like this:-
{{ if isTrue .Env.URL}}

File: tomcat-conf/Catalina/localhost/prweb.xml (hunk 1 above)
Comment (Saurabh-16) [reply tomcat-conf/Catalina/localhost/prweb.xml:38]: Are you referring to prconfig name (`prconfig/Pega-RULES/onPegaK8s`)  or environment variable name (`ON_PEGA_K8S`).

We deinite;y need to pass prconfig/cluster/clientserver/clientmode in prconfig because we want to indicate that pega infinity node is starting in hazelcast client mode.

There are different types of client server model , pega cuttyhunk ,  on-premise and  K8s deployment. So to indicate that this client mode is for K8s deployment , we need to set this value ` <Environment name="prconfig/Pega-RULES/onPegaK8s" value="{{ .Env.ON_PEGA_K8S }}" type="java.lang.String"/>`  to true.

As prconfig/Pega-RULES/onPegaK8s is new setting in prweb.xml and ON_PEGA_K8S is a new environment variable , I feel there should be no problem in changing either of these names.

File: Dockerfile [hunk 4]
```diff
@@ -120,6 +120,13 @@ ENV CASSANDRA_CLUSTER=false \
 # Configure search nodes. Empty string falls back to search being done on the nodes themselves.
 ENV PEGA_SEARCH_URL=
 
+# Configure hazelcast. By default, hazelcast runs in embedded mode.
+Env CLIENT_MODE=false \
```
Comment (pega-embate) [inline Dockerfile:124]: Can we change the name of this env var to something that makes it clear it is related to Hazelcast?  e.g. HZ_CLIENT_MODE?

File: Dockerfile [hunk 5]
```diff
@@ -120,6 +120,13 @@ ENV CASSANDRA_CLUSTER=false \
...
+    ON_PEGA_K8S= \
```
Comment (pega-embate) [inline Dockerfile:125]: Why is this needed?

File: Dockerfile [hunk 6]
```diff
@@ -120,6 +120,13 @@ ENV CASSANDRA_CLUSTER=false \
...
+    CLUSTER_NAME= \
```
Comment (pega-embate) [inline Dockerfile:126]: Actually for all of these env vars please prefix them with HZ to make it clear they are related to hazelcast

File: tomcat-conf/Catalina/localhost/prweb.xml (hunk 1 above)
Comment (pega-embate) [reply tomcat-conf/Catalina/localhost/prweb.xml:38]: What is the behavioral change in Pega when onPegak8s is true?  Is it specific to Hazelcast?

File: Dockerfile (hunk 6 above)
Comment (Saurabh-16) [reply Dockerfile:126]: Sure, I will do this change.

File: Dockerfile (hunk 5 above)
Comment (Saurabh-16) [reply Dockerfile:125]: There are different types of client server model , pega cuttyhunk , on-premise and K8s deployment. So to indicate that this client mode is for K8s deployment , we need to set this value <Environment name="prconfig/Pega-RULES/onPegaK8s" value="{{ .Env.ON_PEGA_K8S }}" type="java.lang.String"/> to true.

**Comment by Emilio-Pega** [comment]
Quick question...

Are al t hese hazelcast changes backwards-compatible or is this going to require downstream changes to the pega, pega-packedengine, and various sde images we build and use?

File: tomcat-conf/Catalina/localhost/prweb.xml (hunk 1 above)
Comment (Emilio-Pega) [reply tomcat-conf/Catalina/localhost/prweb.xml:38]: > Why do we need to have a parameter that specifies architecture implementation? Should we be naming this parameter based on the action it performs instead?
> 
> Also, this Docker image is only used on Kubernetes deployments today.

Actually, we use this as the base image for running local instances for pega developers, so I'm worried all this new hazelcast stuff might break what we have for local dev

Comment (Saurabh-16) [reply tomcat-conf/Catalina/localhost/prweb.xml:38]: It should not break. The default behavior in this image is to run hazelcast in embedded mode which happens today also.
HZ client server work is still WIP. The change we have done in this image is passing few settings in prweb.xml and then added few additional environment variables in hazelcast. They should not break any current deployment process.

File: Dockerfile (hunk 4 above)
Comment (Saurabh-16) [reply Dockerfile:124]: Sure.

**Comment by Saurabh-16** [comment]
> Quick question...
> 
> Are all these hazelcast changes backwards-compatible or is this going to require downstream changes to the pega, pega-packedengine, and various sde images we build and use?

These are backwards-compatible . The platform changes supports both embedded and client-server mode based on settings passed in prweb.xml .

In the docker image code , by default , we are running it in embedded mode and hence we should not see any downstream changes to any of the dependent image.  However , when we switch to client -server mode , then we may need to pass few extra arguments to container to enable hz cs mode.

**Comment by viper0131** [comment]
Why put these settings in here?  Keep the base pure and simple. Advanced settings can be done by overriding the base image.

File: tomcat-conf/Catalina/localhost/prweb.xml (hunk 1 above)
Comment (SushmaReddyLoka) [reply tomcat-conf/Catalina/localhost/prweb.xml:38]: > What is the behavioral change in Pega when onPegak8s is true? Is it specific to Hazelcast?

yes its specific to hazelcast. We need this flag to determine the discovery strategy to be used. On aws , infinity makes use of a customized discovery plugin for discovering servers , in embedded we make use of tcp/ip and for kubernetes we make use of kubernetes hazelcast plugin.

File: Dockerfile (hunk 5 above)
Comment (PraneethaYanala) [reply Dockerfile:125]: on_pega_k8s is used by hazelcast cluster module, if this config is true then it will use kubernetes based discovery between hazelcast clients and servers. We have cuttyhunk based client-server model as well so we are using this setting to differentiate between aws based discovery and kubernetes based discovery for hazelcast

Comment (pega-embate) [reply Dockerfile]: In that case, can we change the name to reflect what it is used for?  e.g. HZ_DISCOVERY_K8S

**Comment by Saurabh-16** [comment]
@Emilio-Pega @dcasavant @pega-embate : - I and Data Grid replied to your queries. I pushed few commits to address code review comments. Please check and let us know if we can still improve this PR.

By default `HZ_CLIENT_MODE` is set to false to notify infinity to run in embedded mode. So I feel its ok to check for this variable in if condition to ensure whether we need to pass hazelcast specific settings to infinity or not.

File: Dockerfile (hunk 5 above)
Comment (Saurabh-16) [reply Dockerfile]: Sure , I will rename this parameter to `HZ_DISCOVERY_K8S`

File: tomcat-conf/Catalina/localhost/prweb.xml (hunk 1 above)
Comment (dcasavant) [reply tomcat-conf/Catalina/localhost/prweb.xml]: If this is Hazelcast specific, it needs a different name that ties it to the *behavior* that is impacted. For example, if this is for connecting to the docker based Hz image, you should name the flag for that.

File: tomcat-conf/Catalina/localhost/prweb.xml (hunk 2 above)
Comment (dcasavant) [reply tomcat-conf/Catalina/localhost/prweb.xml]: A single URL won't change.  And if the DNS lookup changes, the deployment config would need to anyways.  By specifying a namespace, it is tied to the implementation of Kubernetes rather than just container orchestration.

File: tomcat-conf/Catalina/localhost/prweb.xml [hunk 7]
```diff
@@ -32,5 +32,12 @@
   <Environment name="prconfig/session/ha/enabled" value="true" type="java.lang.String"/>
   <Environment name="prconfig/dsm/services/stream/pyBaseLogPath" value="/opt/pega/kafkadata" type="java.lang.String" />
   <Environment name="prconfig/cloud/isNodeAgnosticAdminStudio" value="true" type="java.lang.String"/>
+
+  {{ if isTrue .Env.HZ_CLIENT_MODE }}
+  <Environment name="prconfig/cluster/clientserver/clientmode" value="{{ .Env.HZ_CLIENT_MODE }}" type="java.lang.String"/>
+  <Environment name="prconfig/cluster/clientserver/hzOnK8s" value="{{ .Env.HZ_DISCOVERY_K8S }}" type="java.lang.String"/>
+  <Environment name="prconfig/identification/cluster/name" value="{{ .Env.HZ_CLUSTER_NAME }}" type="java.lang.String"/>
+  <Environment name="prconfig/cluster/clientserver/server/dns" value="{{ .Env.HZ_DNS }}" type="java.lang.String"/>
```
Comment (dcasavant) [inline tomcat-conf/Catalina/localhost/prweb.xml:40]: Do you really need the address of the DNS server?  Or is this the URL of the Hazelcast server?

Comment (pega-embate) [reply tomcat-conf/Catalina/localhost/prweb.xml:40]: I'd also like to understand what this HZ_DNS env var is used for.

Comment (Saurabh-16) [reply tomcat-conf/Catalina/localhost/prweb.xml]: We have decided to go with this name `HZ_SERVER_URL` .
It will be used for kubernetes based discovery between hazelcast clients and servers.
https://github.com/hazelcast/hazelcast-kubernetes#dns-lookup

File: tomcat-conf/Catalina/localhost/prweb.xml (hunk 1 above)
Comment (SushmaReddyLoka) [reply tomcat-conf/Catalina/localhost/prweb.xml]: Yes @dcasavant , we renamed the config to "cluster/clientserver/hzOnK8s".

**Review by kishorv10 — APPROVED** [review]

**Review by MadhuriArugula — APPROVED** [review]
